### PR TITLE
Avatar Features

### PR DESCRIPTION
**Pull request by abuneri**
A few feature additions related to avatars:

- [x] configure avatar data to be read in from stream when user wants to update their avatar

current usage:
```cpp
using namespace concurrency::streams;

//file_stream comes included from disccord/rest/api_client.hpp
//using unsigned char stream bcuz the base64 conversion method takes a std::vector<unsigned char>
auto image_stream = file_stream<unsigned char>::open_istream("path/to/file.jpg").get();
modify_current_user_args args;
args.set_avatar(image_stream);
```
current thoughts:
- want to play with this impl a little bit, open to suggestions

- [x] add a method to `models::user` that gets the avatar URL pre-built based on hash and user id.
```cpp
//something like this
auto usr = client.get_current_user().get();
create_message_args args(usr.get_avatar_url().get_value());
client.create_message(<chan_id>, args);
```
current thoughts:
- could let user specify format (jpg/png) and size, but for now will just default

UPDATE: Both additions have been tested and are working fine

### CODE REVIEW

**Comment by abuneri** [comment]
@FiniteReality here are some quick impl's, any input?

File: include/disccord/models/user.hpp [hunk 1]
```diff
@@ -28,7 +29,7 @@ namespace disccord
 
             private:
                 std::string username;
-                util::optional<std::string> avatar, email;
+                util::optional<std::string> avatar, email, avatar_url;
```
Comment (FiniteReality) [inline include/disccord/models/user.hpp]: Instead of declaring a new field, define `get_avatar_url`

File: lib/models/user.cpp [hunk 2]
```diff
@@ -40,6 +40,14 @@ namespace disccord
             get_field(mfa_enabled, as_bool);
             get_field(verified, as_bool);
             get_field(email, as_string);
+            
+            if (get_avatar().is_specified())
```
Comment (FiniteReality) [inline lib/models/user.cpp]: Likewise, should use a method to create a url instead

File: lib/rest/models/modify_current_user_args.cpp [hunk 3]
```diff
@@ -30,7 +30,17 @@ namespace disccord
...
 
             define_set_method(name, std::string)
-            define_set_method(avatar, std::string)
+            
+            void modify_current_user_args::set_avatar(concurrency::streams::basic_istream<unsigned char> avatar_stream)
+            {
+                concurrency::streams::container_buffer<std::vector<unsigned char>> stream_buffer;
+                avatar_stream.read_to_end(stream_buffer).get();
+                auto stream_bytes = std::move(stream_buffer.collection());
+                avatar_stream.close();
```
Comment (FiniteReality) [inline lib/rest/models/modify_current_user_args.cpp]: Do not close the stream the user passes in, it is not our responsibility.

**Review by FiniteReality — CHANGES_REQUESTED** [review]
Most of the changes look good to me, though I'd prefer it if the avatar_url was not pre-constructed on model decode. There are reasons for this, which I'll demonstrate after I merge the PR.

**Review by FiniteReality — APPROVED** [review]

**Pull request merged**